common/thrift_router.h

```diff
@@ -385,34 +383,47 @@ class ThriftRouter {
             v.push_back(hi.first);
           }
         }
-        MoveUpLocalAndRotateRemote(&v, rotation_counter);
+        RankHostsByGroupPrefixLength(&v, rotation_counter, segment);
       }
       return v;
     }
 
-    void MoveUpLocalAndRotateRemote(
+    void RankHostsByGroupPrefixLength(
         std::vector<const Host*>* v,
-        const unsigned rotation_counter) {
-      auto comparator = [this] (const Host* h1, const Host* h2) -> bool {
-        if (h1->az != h2->az &&
-            (h1->az == this->local_az_ || h2->az == this->local_az_)) {
-          return h1->az == this->local_az_;
-        } else {
-          return h1->az < h2->az;
-        }
+        const unsigned rotation_counter,
+        const std::string& segment) {
+      if (v->size() == 0) return;
+      auto comparator = [this, segment]
+        (const Host* h1, const Host* h2) -> bool {
+        // Descending order
```

Comment:
How about we do something like this, then we can remove all the code after line 402 in this function.

if (h1->groups_prefix_lengths[segment] != h2->groups_prefix_lengths[segment]) {
  return *** > ***;
}

return (rotation_counter & 0x1) ? h1 < h2 : h2 < h1; // this is probably good enough, if it turns out to be an issue, we can add pid info to it.

Reply:
not sure how does it work?

Reply:
What we need to do is to sort hosts by their common prefix length. If some hosts are of the same length, we want to sort them in a different order for each function call.

The way I suggested will sort hosts of the same length based on their address either in descending or ascending order. This will work unless we have more than 2 hosts of the same length AND the hosts' addresses are in the same order in every process in a cluster. If this does happen, we can incorporate pid info into the sorting criteria. But we don't bother to do it now.

Reply:
this implementation has a problem. For example we have 3 hosts both have the same length, and we call getClientsFor() to get ONE host. Because the rotation_counter only has 2 states (&0x1), we will only end up getting 2 orders: either 123 or 321. 

Then the second host will never get called, which made the load not balanced.

Reply:
Did you really read my last comment? I feel the communication is not effective.

Reply:
yeah, i did. here is the comparator i wrote:

auto comparator = [this, &segment, &rotation_counter]
        (const Host* h1, const Host* h2) -> bool {
        // Descending order
        auto s1 = h1->groups_prefix_lengths.at(segment);
        auto s2 = h2->groups_prefix_lengths.at(segment);
        if (s1 == s2) {
          return (rotation_counter & 0x1) ? h1 > h2 : h2 > h1;
        } else {
          return s1 > s2;
        }
      };